synphot/observation.py

```diff
@@ -694,3 +694,220 @@ def as_spectrum(self, binned=True, wavelengths=None):
+        Number of pixels used in the background estimation with units of
+        pixels. Default is set to np.inf * astropy.units.pixel
+        such that there is no contribution of error due to background
+        estimation. This assumes that n_background will be >> npix.
+    background_rate : `~astropy.units.Quantity`, optional
+        Photons per pixel per second due to the backround/sky with units of
+        electrons/second/pixel.
+        Default is 0 * (astropy.units.electron /
+        astropy.units.second / astropy.units.pixel)
+    darkcurrent_rate : `~astropy.units.Quantity`, optional
+        Electrons per pixel per second due to the dark current with units
+        of electrons/second/pixel.
+        Default is 0 * (astropy.units.electron /
+        astropy.units.second / astropy.units.pixel)
+    readnoise : `~astropy.units.Quantity`, optional
+        Electrons per pixel from the read noise with units of electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    gain : `~astropy.units.Quantity`, optional
+        Gain of the CCD with units of electrons/ADU. Default is
+        1 * astropy.units.electron / astropy.units.adu such that the
+        contribution to the error due to the gain is assumed to be small.
+    ad_err : `~astropy.units.Quantity`, optional
+        An estimate of the 1 sigma error within the A/D converter with units of
+        adu/pixel. Default is set to
+        sqrt(0.289) * astropy.units.adu / astropy.units.pixel, where
+        sqrt(0.289) comes from the assumption that "for a charge level that is
+        half way between two output ADU steps, there is an equal chance that it
+        will be assigned to the lower or to the higher ADU value when converted
+        to a digital number" (text from footnote on page 56 of [1]_, see also
+        [2]_).
+
+    References
+    ----------
+    .. [1] Howell, S. B. 2000, *Handbook of CCD Astronomy* (Cambridge, UK:
+        Cambridge University Press)
+    .. [2] Merline, W. & Howell, S. B. *A Realistic Model for Point-sources
+        Imaged on Array Detectors: The Model and Initial Results*. ExA, 6:163
+        (1995)
+
+    Returns
+    -------
+    t : `~astropy.units.Quantity`
+        The exposure time needed (in seconds) to achieve the given signal
+        to noise ratio.
+    """
+    readnoise = _get_shotnoise(readnoise)
+    gain_err = _get_shotnoise(gain * ad_err)
+
+    # solve t with the quadratic equation (pg. 57 of Howell 2000)
+    A = countrate ** 2
+    B = (-1) * snr ** 2 * (countrate + npix * (background_rate +
+                                               darkcurrent_rate))
+    C = (-1) * snr ** 2 * npix * readnoise ** 2
+
+    t = (-B + np.sqrt(B ** 2 - 4 * A * C)) / (2 * A)
+
+    if gain_err.value > 1 or not np.isfinite(n_background.value):
+        from scipy.optimize import fsolve
+        # solve t numerically
+        t = fsolve(_t_with_small_errs, t, args=(background_rate,
```

Comment:
I'm confused by why `howell_snr` is not called in this function (and its use in `exptime_from_howell_snr`)? It would definitely be better to have a shared implementation between the two...

Reply:
Hm, maybe the `from_howell_snr` bit in `exptime_from_howell_snr` is misleading, because the two snr expressions are slightly different - `howell_snr` assumes you already know the exposure time (i.e. you input `counts` instead of `countrate`, `darkcurrent` instead of `darkcurrent_rate`, etc), where the expression for the SNR in `exptime_from_howell_snr` is solving for `t` assuming you only know the count rates of each component.

Let me know if that doesn't answer the question and I'll give it another shot!

Reply:
I guess I named it `exptime_from_howell_snr` instead of `exptime_from_snr` to be explicit that the expression comes from the CCD Handbook (and not just something I made up, hah), though I do see how the name leads one to think that this function uses the `howell_snr` function to get `t`. I'm open to suggestions for a different name!

Reply:
https://twitter.com/codinghorror/status/506010907021828096?lang=en -- Erik, you got some name suggestions?

Reply:
> the two SNR expressions are slightly different 

Ah, I see your point here.  It *feels* like it should be possible to unify these a bit, but it's not obvious to me how and the perfect is the enemy of the good!